ansible/roles-infra/infra-ec2-template-generate/templates/cloud_template.j2

```diff
@@ -149,133 +151,135 @@ Resources:
             "Fn::GetAtt":
               - DnsZonePublic
               - NameServers
-    {% endif %}
+{% endif %}
 
 {% for instance in instances %}
-{% if instance['dns_loadbalancer'] | d(false) | bool
-  and not instance['unique'] | d(false) | bool %}
+{%   if instance['dns_loadbalancer'] | default(false) | bool
+     and not instance['unique'] | default(false) | bool %}
   {{instance['name']}}DnsLoadBalancer:
     Type: "AWS::Route53::RecordSetGroup"
     DependsOn:
-    {% for c in range(1, (instance['count']|int)+1) %}
+{%     for c in range(1, (instance['count']|int)+1) %}
       - {{instance['name']}}{{c}}
-      {% if instance['public_dns'] %}
+{%       if instance['public_dns'] %}
       - {{instance['name']}}{{c}}EIP
-      {% endif %}
-    {% endfor %}
+{%       endif %}
+{%     endfor %}
     Properties:
-      {% if secondary_stack is defined %}
+{%     if not aws_dns_create_public_zone | bool %}
+      HostedZoneName: "{{ aws_dns_zone_root }}"
+{%     elif secondary_stack is defined %}
       HostedZoneName: "{{ aws_dns_zone_public }}"
-      {% else %}
+{%     else %}
       HostedZoneId:
         Ref: DnsZonePublic
-      {% endif %}
+{%     endif %}
```

Comment:
i'm not sure to understand the logic here.
What happens if `secondary_stack` is true and `aws_dns_create_public_zone` is true ?

It would try to use the ref DnsZonePublic which should not be present in the current stack (as it is secondary).

I think you want:

```jinja2
{%     if aws_dns_create_public_zone | bool %}
{%       if secondary_stack is defined %}
      HostedZoneName: "{{ aws_dns_zone_public }}"
{%       else %}
      HostedZoneId:
        Ref: DnsZonePublic
{%       endif %}
{%     else %}
{%       if secondary_stack is defined %}
      HostedZoneName: "{{ aws_dns_zone_public }}"
{%       else %}
      HostedZoneName: "{{ aws_dns_zone_root }}"
{%     endif %}
```

Reply:
Good point. I changed it to, which I believe is logically equivalent to what you suggested:
```
{%     if secondary_stack is defined %}
      HostedZoneName: "{{ aws_dns_zone_public }}"
{%     elif not aws_dns_create_public_zone | bool %}
      HostedZoneName: "{{ aws_dns_zone_root }}"
{%     else %}
      HostedZoneId:
        Ref: DnsZonePublic
{%     endif %}
```

Reply:
Not quite.

In your code snippet, if `secondary_stack` is true and `aws_dns_create_public_zone` is false, your template will still use

`aws_dns_zone_public` and not `aws_dns_zone_root`

You want both conditions nested IMO.